### PR TITLE
Add BaseFilesystemReader interface + implement in SimpleDirectoryReader

### PR DESCRIPTION
**Pull request by Javtor**
# Description

This is just the changes made to `llama-index-core` in #13408. Opened a separate PR for these because the changes done to integrations there (like S3 and SharePoint) depend on the core package version being bumped first to pass CI.

#13408 also contains the unit tests for the implementation in SimpleDirectoryReader

## Version Bump?
- [x] Yes

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: llama-index-core/llama_index/core/readers/file/base.py
Comment (sourabhdesai) [inline llama-index-core/llama_index/core/readers/file/base.py:34]: Am wondering if this should return a `FileInfo`/`FileMetadata` pydantic class which just contains some minimum required metadata fields to standardize this a bit 🤔

File: llama-index-core/llama_index/core/readers/file/base.py
Comment (sourabhdesai) [inline llama-index-core/llama_index/core/readers/file/base.py:43]: should this return the file content instead? to separate the file parsing from file loading

**Review by sourabhdesai — COMMENTED** [review]
mainly have a question here about whether the `read_file` methods should actually be returning the file content 🤔 e.g. `bytes`?

File: llama-index-core/llama_index/core/readers/base.py
Comment (logan-markewich) [inline llama-index-core/llama_index/core/readers/base.py:74]: This looks good to me, and probably fine to merge for now.

But in general, nesting base class under base class can be a confusing pattern to maintain 😅 Longer term I'd prefer just a single base class, or mixin's that add functionality 

Because now we have BaseReader -> BaseResourcesReader -> BaseFileSystemReader -> <Actual reader implemention>

File: llama-index-core/pyproject.toml
Comment (logan-markewich) [inline llama-index-core/pyproject.toml:46]: fyi this would need to be manually published after merging (happy to do that)

File: llama-index-core/llama_index/core/readers/file/base.py
Comment (logan-markewich) [inline llama-index-core/llama_index/core/readers/file/base.py:26]: Kind of curious how this differs from `load_resource()` -- feels like it would be the same method?

**Review by logan-markewich — APPROVED** [review]
Approving for now to unblock work, but
- there are too many base classes
- BaseFilesSystemReader doesn't seem entirely needed? But I could be wrong

File: llama-index-core/llama_index/core/readers/file/base.py
Comment (logan-markewich) [reply llama-index-core/llama_index/core/readers/file/base.py:26]: I guess this returns bytes

**Pull request merged**
